subsys/net/openthread/rpc/server/ot_rpc_meshdiag.c

```diff
@@ -0,0 +1,211 @@
+		ot_rpc_report_cmd_decoding_error(OT_RPC_CMD_MESH_DIAG_GET_NEXT_IP6_ADDRESS);
+		return;
+	}
+
+	openthread_api_mutex_lock(openthread_get_default_context());
+	error = otMeshDiagGetNextIp6Address(iterator, &address);
+	openthread_api_mutex_unlock(openthread_get_default_context());
+
+	if (error != OT_ERROR_NONE) {
+		nrf_rpc_rsp_send_uint(group, error);
+		return;
+	}
+
+	NRF_RPC_CBOR_ALLOC(group, rsp_ctx, 3 + OT_IP6_ADDRESS_SIZE);
+
+	nrf_rpc_encode_uint(&rsp_ctx, error);
+	nrf_rpc_encode_buffer(&rsp_ctx, address.mFields.m8, OT_IP6_ADDRESS_SIZE);
+
+	nrf_rpc_cbor_rsp_no_err(group, &rsp_ctx);
+}
+
+static void ot_rpc_cmd_get_next_child_info(const struct nrf_rpc_group *group,
+					   struct nrf_rpc_cbor_ctx *ctx, void *handler_data)
+{
+	struct nrf_rpc_cbor_ctx rsp_ctx;
+	size_t cbor_buffer_size = 0;
+	otError error;
+	otMeshDiagChildIterator *iterator;
+	ot_rpc_res_tab_key iterator_key;
+	otMeshDiagChildInfo child_info;
+	uint8_t child_flags = 0;
+
+	iterator_key = nrf_rpc_decode_uint(ctx);
+	iterator = ot_res_tab_meshdiag_child_it_get(iterator_key);
+
+	if (!nrf_rpc_decoding_done_and_check(group, ctx)) {
+		ot_rpc_report_cmd_decoding_error(OT_RPC_CMD_MESH_DIAG_GET_NEXT_CHILD_INFO);
+		return;
+	}
+
+	openthread_api_mutex_lock(openthread_get_default_context());
+	error = otMeshDiagGetNextChildInfo(iterator, &child_info);
+	openthread_api_mutex_unlock(openthread_get_default_context());
+
+	if (error != OT_ERROR_NONE) {
+		nrf_rpc_rsp_send_uint(group, error);
+		return;
+	}
+
+	cbor_buffer_size += 2;			  /* aError */
+	cbor_buffer_size += 1 + sizeof(uint16_t); /* mRloc16 */
+	cbor_buffer_size += 1 + sizeof(uint8_t);  /* child_flags */
+	cbor_buffer_size += 1 + sizeof(uint8_t);  /* mLinkQuality */
+
+	NRF_RPC_CBOR_ALLOC(group, rsp_ctx, cbor_buffer_size);
+
+	nrf_rpc_encode_uint(&rsp_ctx, error);
+	nrf_rpc_encode_uint(&rsp_ctx, child_info.mRloc16);
+
+	WRITE_BIT(child_flags, 0, child_info.mMode.mRxOnWhenIdle);
+	WRITE_BIT(child_flags, 1, child_info.mMode.mDeviceType);
+	WRITE_BIT(child_flags, 2, child_info.mMode.mNetworkData);
+	WRITE_BIT(child_flags, 3, child_info.mIsThisDevice);
+	WRITE_BIT(child_flags, 4, child_info.mIsBorderRouter);
+
+	nrf_rpc_encode_uint(&rsp_ctx, child_flags);
+	nrf_rpc_encode_uint(&rsp_ctx, child_info.mLinkQuality);
+
+	nrf_rpc_cbor_rsp_no_err(group, &rsp_ctx);
+}
+
+static void handle_mesh_diag_discover(otError aError, otMeshDiagRouterInfo *aRouterInfo,
+				      void *aContext)
+{
+	struct nrf_rpc_cbor_ctx ctx;
+	size_t cbor_buffer_size = 0;
+	uint8_t router_flags = 0;
+	ot_rpc_res_tab_key child_iterator_key;
+	ot_rpc_res_tab_key ip6_iterator_key;
+
+	ARG_UNUSED(aContext);
+
+	ip6_iterator_key = ot_res_tab_meshdiag_ip6_it_alloc(aRouterInfo->mIp6AddrIterator);
+	child_iterator_key = ot_res_tab_meshdiag_child_it_alloc(aRouterInfo->mChildIterator);
+
+	if ((aError == OT_ERROR_NONE || aError == OT_ERROR_PENDING) &&
+	    (ip6_iterator_key == 0 || child_iterator_key == 0)) {
+		aError = OT_ERROR_NO_BUFS;
```

Comment:
This is undocumented error code for `otMeshDiagDiscoverCallback`, so probably shouldn't be used. Btw, we don't expect this to be possible in practice, right? So maybe `nrf_rpc_err` should be enough?

Btw. I think in some cases OT may return NULL iterators - if discovery of children and address is not enabled. In such a case, it is expected that ot_res_tab_meshdiag_ip6_it_alloc/ot_res_tab_meshdiag_child_it_alloc returns NULL,right?

Reply:
yeah, you are right. I've changed the code to raise error only if pointer is not null, but result of allocation is equal to 0